File: leetcode/1137_nth_tribonacci_number.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <cassert>

using namespace std;

class Solution {
 public:
    int tribonacci(int n) {
        if (n == 0) {
            return 0;
        } else if (n == 1) {
            return 1;
        } else if (n == 2) {
            return 1;
        } else {
            int x = 0, y = 1, z = 1;

            for (int i = 3; i <= n; i++) {
                int tmp = x + y + z;
                x = y;
                y = z;
                z = tmp;
            }

            return z;
        }
    }
};

int main() {
    auto solution = Solution();

    assert(4 == solution.tribonacci(4));
    assert(1389537 == solution.tribonacci(25));
}

File: leetcode/1137_nth_tribonacci_number.py
# -*- coding: utf-8 -*-


class Solution:
    def tribonacci(self, n):
        if n == 0:
            return 0
        elif n == 1:
            return 1
        elif n == 2:
            return 1
        else:
            x, y, z = 0, 1, 1
            for _ in range(3, n + 1):
                x, y, z = y, z, x + y + z
            return z


if __name__ == '__main__':
    solution = Solution()

    assert 4 == solution.tribonacci(4)
    assert 1389537 == solution.tribonacci(25)
